bin/demo.py

```diff
@@ -63,15 +66,49 @@ def stdchannel_redirected(stdchannel, dest_filename):
         if dest_file is not None:
             dest_file.close()
 
-def check_and_set_token():
+def needs_authentication():
+    token = dcos.config.get_config_val("core.dcos_acs_token")
+    if token is None:
+        # need to check if token is set because the CLI code will prompt for auth otherwise
+        return True
+    else:
+        try:
+            # make a request that requires authentication
+            shakedown.dcos_leader()
+            return False
+        except dcos.errors.DCOSException:
+            return True
+
+def authenticate_with_oauth(dcos_url):
+    dcos_oauth_token = arguments['--dcos-oauth-token']
+    url = dcos_url + 'acs/api/v1/auth/login'
+    creds = { 'token': dcos_oauth_token }
+    r = http.request('post', url, json=creds)
+    if r.status_code == 200:
+        json_data = r.json()
+        if 'token' not in json_data:
+            log_and_exit('!! token not returned from authentication request; is DC/OS healthy?')
+        dcos.config.set_val('core.dcos_acs_token', json_data['token'])
+    else:
+        log_and_exit('!! DC/OS authentication failed; ' +
+            'invalid --dcos-oauth-token provided')
+
+def authenticate_with_username():
     dcos_username = arguments['--dcos-username']
     dcos_password = arguments['--dcos-password']
-    try:
-        token = shakedown.authenticate(dcos_username, dcos_password)
-        dcos.config.set_val('core.dcos_acs_token', token)
-    except:
-        log_and_exit('!! DC/OS authentication failed; ' +
-            'did you provide --dcos-username and --dcos-password?')
+    token = shakedown.authenticate(dcos_username, dcos_password)
+    dcos.config.set_val('core.dcos_acs_token', token)
+
+def check_and_set_token(dcos_url):
+    if needs_authentication():
+        dcos_oauth_token = arguments['--dcos-oauth-token']
+        if dcos_oauth_token:
+            authenticate_with_oauth(dcos_url, dcos_oauth_token)
+        else:
+            authenticate_with_username()
```

Comment:
Previously, this was surrounded by `try/catch`. Are we ok with an exception getting thrown here? (I think it will exit with some info for the user)